build_runner/lib/src/generate/build.dart

```diff
@@ -159,46 +149,42 @@ Stream<BuildResult> serve(List<BuildAction> buildActions,
       directory: directory,
       address: address,
       port: port);
-  var watchImpl = new WatchImpl(options, buildActions);
+  var terminator = new _Terminator(terminateEventStream);
+  var buildResults =
+      runWatch(options, buildActions, terminator.shouldTerminate);
 
-  var resultStream = watchImpl.runWatch();
-  var serverStarted = startServer(watchImpl, options);
+  var serverStarted = startServer(buildResults, options);
 
-  // Stop doing new builds when told to terminate.
-  _setupTerminateLogic(terminateEventStream, () async {
-    await watchImpl.terminate();
-  }, cancelWhen: watchImpl.onTerminated);
-
-  watchImpl.onTerminated.then((_) async {
+  buildResults.builds.drain().then((_) async {
+    await terminator.cancel();
     await serverStarted;
     await stopServer();
     await options.logListener.cancel();
   });
 
-  return resultStream;
+  return buildResults.builds;
 }
 
-/// Given [terminateEventStream], call [onTerminate] the first time an event is
-/// seen. If a second event is received, simply exit.
-StreamSubscription _setupTerminateLogic(
-    Stream terminateEventStream, Future onTerminate(),
-    {Future cancelWhen}) {
-  terminateEventStream ??= ProcessSignal.SIGINT.watch();
-  int numEventsSeen = 0;
-  StreamSubscription terminateListener;
-  terminateListener = terminateEventStream.listen((_) {
-    numEventsSeen++;
-    if (numEventsSeen == 1) {
-      onTerminate().then((_) {
-        terminateListener.cancel();
-      });
-    } else {
-      exit(2);
-    }
-  });
-
-  cancelWhen?.then((_) {
-    terminateListener.cancel();
-  });
-  return terminateListener;
+class _Terminator {
+  final Future shouldTerminate;
+  final StreamSubscription _subscription;
+
+  factory _Terminator(Stream terminateEventStream) {
```

Comment:
nit: if this supports null values it should probably be an optional arg (if for no other reason than to send the right signal that it isn't required)

Reply:
Done